### PR TITLE
[flang] Implement legacy %VAL and %REF actual arguments

### PR DESCRIPTION
**Pull request by jeanPerier**
Update evaluate::ActualArgument to propagate the %VAL and %REF markers until lowering.
Semantic checks are added to %VAL to ensure the argument is a numerical or logical scalar.

I did not push these markers into the characteristics because other compilers do not complain about inconsistent usages (e.g. using %VAL in a call on a procedure with an interface without VALUE dummies is not flagged by any compilers I tested, and it is not an issue for lowering, so I decided to stay simple here and minimize the footprint of these legacy features).

Lowering retrieve these markers and does the right thing: pass %VAL in registers and pass %REF by address without adding any extra arguments for characters.

Note that %LOC was already handled (rewritten as LOC intrinsic), but this patch tests that %VAL(%LOC()) allows skipping copy-in/copy-out as described [in gfortran documentation](https://gcc.gnu.org/onlinedocs/gcc-3.4.6/g77/_0025LOC_0028_0029.html#g_t_0025LOC_0028_0029).

### CODE REVIEW

**Review by razvanlupusoru — COMMENTED** [review]
Looks great!

**Review by clementval — COMMENTED** [review]
LGTM. I'll let Peter give the final approval.

File: flang/lib/Semantics/expression.cpp [hunk 1]
```diff
@@ -3898,32 +3900,52 @@ void ArgumentAnalyzer::Analyze(const parser::Variable &x) {
...
+              actual->set_isPercentRef();
+            }
+          },
+          [&](const parser::ActualArg::PercentVal &percentVal) {
+            actual = AnalyzeExpr(percentVal.v);
+            if (actual.has_value()) {
+              actual->set_isPercentVal();
+              std::optional<DynamicType> type{actual->GetType()};
+              if (!type ||
+                  !(common::IsNumericTypeCategory(type->category()) ||
```
Comment (klausler) [inline flang/lib/Semantics/expression.cpp:3940]: `evaluate::IsLengthlessIntrinsicType()` can apply here.

**Review by klausler — APPROVED** [review]
Very well done; thank you!

File: flang/lib/Semantics/expression.cpp (hunk 1 above)
Comment (jeanPerier) [reply flang/lib/Semantics/expression.cpp]: Thanks for the pointer, [the one I found ](https://github.com/llvm/llvm-project/blob/1914bcfd3fb4d37b154e9cfeddcec3c04766d1b7/flang/include/flang/Evaluate/type.h#L382) operates on template arguments, so I added an equivalent for DynamicType.

**Pull request merged**